### DISCUSSION TITLE
Is it possible to nest instrument channels for windows, traces, and markers for a given channel?

### DISCUSSION THREAD

**Discussion by Sionwage** [pymeasure/pymeasure#1072] [Q&A]
While writing up a VNA class instrument, I am looking at how to handle multiple traces, windows, and markers.

For a Keysight 5071 two port VNA, you have 16 channels available for rapidly switching between different configurations. Each channel has upto 16 unique traces that could be S-parameter measurements or absolute measurements that take the ratio between a port or two ports and displayed in various formats like using Logmag, Polar Plot, Smith Chart, phase, etc.

Shared between the up-to 16 traces for a channel are 8 markers.

Is it possible to nest channels within channels to account for both a channel and a trace? I am going to experiment to see if I can but I figured it wouldn't hurt to ask and try to explain what I am trying to do.

One instance is a SCPI command `CALC{ch}:PAR{tr}:DEF S12` to set trace of a channel to measure S12.

It looks like it might be possible to assign trace 1-16 to channels 1-16, or at least define how many channels and how many traces are to be used when initializing a `Instrument` so that I could chain the trace to the channel like this `vna.ch_1.trace_1.measurement_parameter = 'S12'` as long as I initialize the traces in the channel class with something other than `'ch_` as the placeholder.

Another example would be reading the value of multiple markers. If I have a phase plot and I want to read the phase of marker 3 on trace 5: `print(vna.ch_1.trace_5.marker_3.value)` or `print(vna.ch_1.marker_3.value)`. The SCPI command in this instance would be `CALC{ch}:MARK3:X?` which simply reads the currently active trace for the channel. I am trying to avoid using tuples to pass which marker to select and what frequency to to move it to for readability.

Another case is I have to first send a command to declare how many traces are used for the channel before selecting a trace to define its parameters by sending `CALC{ch}:PAR:COUN {1-16}` for the number of traces on that channel.

It would be nice to be able to chain the properties of the instrument so the active channel and the active trace could be kept track of to switch to the appropriate channel, trace, marker, etc or issue the necessary commands to allow use when needed but not slow down the communication by sending them when not needed.

I'll experiment with nesting the `MultiChannelCreator` function within Channel classes.

I think I can use the `get_process` and `set_process` to intercept the selected trace, channel, marker, or whatever value to identify if I need to issue commands first before completing the command to ensure it executes successfully such as making a trace active.

I would like to dynamically update the channels, traces, markers available as instrument properties but I think its not possible with how the library is written to prevent overwriting instrument properties.

**Comment by BenediktBurger** [comment]
You can add `Channels` to `Channels`, as that was exactly one goal of the Channel implementation. For that reason, a channel has a `parent`, as parameter, not the `instrument`.

@CasperSchippers added a vector analyzer (Anritsu MS4644B) in #827 which has also traces and channels and markers. It also has some update routine for channel numbers etc. Maybe you could take a look at that instrument.

**Reply by Sionwage** [reply]
Thank you!

How to handle the `format_map` string function with `SENS{ch}:PAR{tr}:SEL` was answered by the code how he nested the ch and tr together using double braces for the outer channel like this `":CALC{{ch}}:PAR{tr}:DEF?"`

I gotta figure out if I can clean this up for turning a list of numbers into a pair used for complex numbers:

    measurement_data = Channel.measurement(
        "CALC{ch}:DATA:SDAT?",  # read out real/imag data for active trace with corrections
        """ """,
        cast=complex,
        preprocess_reply=lambda x: ",".join(
            [f"{float(i)}+{float(j)}j" for i, j in zip(x.split(",")[0::2], x.split(",")[1::2])],
        ),
    )

Otherwise if I cast it as complex it is the same list with +0j appended to each number.

**Reply by Sionwage** [reply]
This is how the data is coming back

    '+1.00526733920E+000,+2.15773174324E-003,+6.21223234330E-001,+7.86645668658E-001,-2.40556185956E-001,+9.82023805881E-001,...'

**Reply by BenediktBurger** [reply]
I'd keep the `cast=float` (default setting), but use `get_process = lambda vals: [vals[i] + vals[i + 1] * 1j for i in range(0, len(vals), 2)]` to turn the list of floats into a list of complex.

btw: the double curly braces work, because `str.format` replaces double curly braces with a single version and only does evaluate the content for single braces. Therefore you can nest it several times with using several curly braces.
